frame/nomination-pools/src/lib.rs

```diff
@@ -1550,18 +1606,17 @@ pub mod pallet {
 			// payout related stuff: we must claim the payouts, and updated recorded payout data
 			// before updating the bonded pool points, similar to that of `join` transaction.
 			reward_pool.update_records(bonded_pool.id, bonded_pool.points)?;
-			// TODO: optimize this to not touch the free balance of `who ` at all in benchmarks.
-			// Currently, bonding rewards is like a batch. In the same PR, also make this function
-			// take a boolean argument that make it either 100% pure (no storage update), or make it
-			// also emit event and do the transfer. #11671
-			let claimed =
+
+			// we have to claim the pending_rewards every time we change the bonded amount.
+			if matches!(extra, BondExtra::FreeBalance(_)) {
 				Self::do_reward_payout(&who, &mut member, &mut bonded_pool, &mut reward_pool)?;
```

Comment:
What we need to avoid all this duplicate code is:

1. break down the current `do_reward_payout` to `calculate_and_payout_rewards`.
2. internally, this is composed of two functions: `calculate_rewards` and `payout_rewards`. 
3. In here, if it is `BondExtra::Rewards`, only calculate, and use the same `try_bond_funds`.
4. If it is `BondExtra::FreeBalance`, use the existing `calculate_and_payout_rewards`.

Reply:
Thanks for specifying me this, I have cleaned up the code, so could you check if the changes I made are reasonable?